src/itox.8
.TH ITOX 8 "March 2005" "xinetd"
.SH NAME
itox \- converts inetd.conf style configuration files to xinetd.conf
.SH SYNOPSIS
itox [\-daemon_dir <dir_name>]
.SH DESCRIPTION
.B itox
takes on its standard input inetd.conf style entries and dumps to
standard output the corresponding xinetd.conf style entries.
.SH OPTIONS
.TP
.I \-daemon_dir <dir_name>
If you use tcpd, this option specifies the directory where all the daemons are.
You must specify this option if you use tcpd and the daemon file names are not
absolute.
.SH EXAMPLES
itox \-daemon_dir /usr/sbin < inetd.conf > xinetd.conf
.SH AUTHOR
xinetd and itox were written by Panagiotis Tsirigotis.
.sp
This man page was written by Norbert Veber <nveber@debian.org> and Thomas Seyrat
<tomasera@debian.org>


CHANGELOG

This file describes the changes to xinetd. The base version is 2.0.0.

2.0.0: (not released)

2.0.1: (not released)

2.0.2:
	Changes to the Makefile.
	Trivial mods to the usage() function

2.0.3:
	Changes to the Makefile.
	Fixed a bug in the usage() function

2.0.4:
	Distribution versions of the Makefile no longer contain any 
	references to options.opt

2.0.5:
	1) Clarified what an "unlisted RPC service" is in the man page
	2) Fixed a bug in remote_address_check which caused access to be denied
		for all hosts if no_access was set and only_from was not set.
	3) Fixed a bug which caused arbitrary syslog levels if the log_type
		was specified as SYSLOG in the "defaults" entry

2.0.6:
	bug fix in child.c: replaced strx_sprint with strx_print when naming
	interceptor processes


-------------------------------------------------------------------------------

2.1.0: (not released)
	This was mostly a clean-up of 2.0
	List of changes:

	A. The man page has been split into 3 parts:
			xinetd.man			:	man page describing the program
			xinetd.conf.man	:	man page describing the configuration file 
										This file now includes figures about the
										overhead of interception.
			xinetd.log.man		:	man page describing the log file
	B. New service attributes
			rpc_number			: to support unlisted RPC services
			nice					: to set the nice value of forked servers
	C. The IDONLY service flag was added
	D. Now uses the timer library unless NO_TIMERS is defined. If NO_TIMERS
		is not defined, the following are also available:
			a) a new option, -cc, to do periodic consistency checks
			b) timeout for reconfiguration
	E.	Configuration file man page now mentions that access control is 
		based on IP-address instead of domain address.
	F.	The interception code now sets the TCP_NODELAY option.
	G. The timeout when contacting a remote identification server is
		configurable both when the service request has been accepted and 
		when the request is rejected. In the former case it defaults to
		infinity while in the latter it defaults to 30 sec.
	H. The log line ids are now constants in an include file.
	I. The fsma library is no longer used.
	J. Most structure fields were renamed; certain functions were renamed too.
		Access to structure fields is now via macros.
	K. A bug in not restoring the number of descriptors to the soft limit
		was fixed.
	L. The -pid option works
	M. New internal services:
			"servers"		: lists active servers
			"services"		: lists active services

2.1.1:
	xinetd produces more meaningful messages if it can't start logging

2.1.2:
	Fixed 2 bugs:
		1) xinetd would request a port number for unlisted RPC services
			One could get around this by simply specifying the port attribute
			for the service; the specified value would be ignored anyway.
		2) xinetd initialization might fail sometimes in function msg_init()
			syslog logging was specified (i.e. the option -syslog was used)
			The cause of this bug was that the 
				xlog_control( xlog, XLOG_GETFD, &fd ) 
			operation will not fail for xlogs connected to syslog and
			the value of 'fd' would be used in the subsequent fcntl(). That
			value is arbitrary since 'fd' is a local variable.
			If 'fd' did not happen to refer to an open descriptor, the program 
			would terminate since the fcntl() would fail (btw, the success
			of the fcntl() call would be harmless if 'fd' happened to
			refer to an open descriptor).

2.1.3:
	Bug fix:
		only_from/no_access addresses would get inverted on little-endian
		machines when such addresses were specified using the numeric notation
		(for example, 128.138.45.3). This bug was in the numeric_addr function
		which did not convert the result of inet_addr to host-byte-order.
		A work around for the bug would be to use the factorized address 
		notation (for example, 128.138.45.{3})

2.1.4:
	Bug fixes:
		1) in dgram_echo(), sin_len was not being set before the invocation of
			recvfrom
		2) in finger_shutdown(), it was possible for Srdline() to return NULL
			(if the remote end would close the socket without sending anything).
			If the RECORD option was set in the log_on_failure flags, this
			would cause the forked xinetd process which did the recording to
			die since it would try to dereference a NULL pointer.


2.1.5:
	Bug fixes:
		1) in exec_server() service descriptor might be closed when
		   execing server. This made it impossible to start servers
			for 'nowait' services.  The bug occurred only Ultrix version
			4.3a or (probably) later (bug discovered and fix provided
			by doug@seas.smu.edu)
		
		2) for systems that supported supplementary group id's, the
			set_credentials() function did not set those group id's
			(they were being inherited from xinetd). Now initgroups(3)
			is called to set the supplementary group id's properly.
			(bug discovered and fix provided by maf+@osu.edu)

2.1.6:
	Bug fixes:
		1) xinetd will crash after reconfiguration if there is a running
			server for a service that was removed from the configuration and
			which logs on exit.
		2) xinetd forked process falls in infinite loop if identd server
			sends a reply that is missing the ending CR-LF.
			(bug discovered and fix provided by Laurent.Wacrenier@gin.obspm.fr).

	We also change the LOGUSER_SUCCESS_TIMEOUT constant in config.h from
	0 seconds (i.e. infinite timeout) to 30 seconds. This avoids infinite
	waits in case the remote host does not send a RST reply when we attempt
	to connect to the IDENTD port, and there is no server listening at
	that port.


2.1.7:
	Bug fix: the HOST flag in the 'log_on_success' attribute was ignored;
				the code was incorrectly checking if the HOST flag was
				set in the log_on_failure attribute
				(bug discovered by frolich@corrine.cpc.cs.ucf.edu)

2.1.8.1: 
	Added support for TCP redirect to a remote host.
		--bbraun

2.1.8.2: 
	Added support for binding to specific interface.
		--bbraun

2.1.8.3:
	Changed redirect so that a remote port can be specified as well.

src/itox.c
/*
 * (c) Copyright 1992 by Panagiotis Tsirigotis
 * (c) Sections Copyright 1998-2001 by Rob Braun
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

#include "config.h"
#define EQ( s1, s2 )				( strcmp( s1, s2 ) == 0 )

#define NUL					'\0'
#define static					static

#define FIELD_WIDTH				15
#define DAEMON_DIR_OPTION			"-daemon_dir"
#define TCPD_NAME				"tcpd"

#include <stdlib.h>
#include <string.h>
#include "sio.h"
#include "str.h"
#if !defined(linux)
#include "misc.h"
#else
#include <libgen.h>
#endif

str_h strp ;
int line_count ;
static void print_line( const char *name, const char *value );
static char *next_word( const char *description );
static char *make_string_cat( register unsigned count, ... );
static char *make_pathname( register unsigned count, ... );


/*
 * This program works only as a filter.
 * Options:
 *    -daemon_dir <dir_name> : if you use tcpd, this option specifies the
 *		directory where all the daemons are.
 *		You must specify this option if you use tcpd
 *
 * Note that we don't bother to free the memory we malloc.
 */
int main(int argc, char *argv[] )
{
	char *s ;
	int uses_tcpd ;
	char *daemon_dirpath = NULL ;

	if ( argc != 1 && argc != 3 )
	{
		Sprint( 2, "Usage: %s [%s dir_path]\n",
				basename( argv[ 0 ] ), DAEMON_DIR_OPTION ) ;
		exit( 1 ) ;
	}

	uses_tcpd = ( argc == 3 ) ;

	if ( uses_tcpd )
	{
		int len ;

		daemon_dirpath = argv[ 2 ] ;
		len = strlen( daemon_dirpath ) ;
		if ( daemon_dirpath[ len-1 ] == '/' )
			daemon_dirpath[ --len ] = NUL ;
	}

	strp = str_parse( (char *)0, " \t", STR_NOFLAGS, (int *)0 ) ;

	while ( (s = Srdline( 0 )) )
	{
		char *word ;
		char *p ;
		char *socket_type, *protocol ;
		char *service ;
		int is_rpc ;

		line_count++ ;

		if ( SIOLINELEN( 0 ) == 0 || s[ 0 ] == '#' )
			continue ;

		str_setstr( strp, s ) ;

		service = word = next_word( "service name" ) ;

		/*
		 * Check if it is an RPC service
		 */
		p = strchr( word, '/' ) ;
		if ( p != NULL )
			*p = 0 ;
		Sprint( 1, "service %s\n{\n", word ) ;
		if ( (is_rpc = ( p != NULL )) )
		{
			print_line( "type", "RPC" ) ;
			print_line( "rpc_version", p+1 ) ;
		}

		socket_type = word = next_word( "socket type" ) ;
		print_line( "socket_type", socket_type ) ;

		word = next_word( "protocol" ) ;
		p = strchr( word, '/' ) ;
		protocol = ( p == NULL ) ? word : p+1 ;

		print_line( "protocol", protocol ) ;

		word = next_word( "wait/nowait" ) ;
		p = strchr(word, '.');
		if (p != NULL)
		{
			Sprint( 2, 
			"The entry for service %s/%s may be wrong, because\n", 
				service, protocol);
			Sprint( 2, 
			"we can't convert .max option for wait/nowait field\n");
			*p = '\0';
			print_line( "wait", EQ( word, "wait" ) ? "yes" : "no" );
		}
		else
			print_line( "wait", EQ( word, "wait" ) ? "yes" : "no" );

		word = next_word( "user[.group]" ) ;
		p = strchr(word, '.');
		if (p != NULL)
		{
			*p = '\0';
			print_line( "user", word ) ;
			word = ++p;
			print_line( "group", word );
		}
		else
			print_line( "user", word ) ;

		word = next_word( "server" ) ;
		if ( EQ( word, "internal" ) )
		{
			/*
			 * We are in trouble if this is an RPC service
			 */
			if ( is_rpc )
			{
				Sprint( 2,
					"The entry for service %s will be wrong because\n", service ) ;
				Sprint( 2, "we can't handle internal RPC services\n" ) ;
			}
			else
			{
				print_line( "type", "INTERNAL" ) ;
				print_line( "id", 
					make_string_cat( 3, service, 
						"-", socket_type ) ) ;
			}
		}
		else
		{
			char *server_path = word ;	/* from inetd.conf */
			char *server_of_server_path = basename( server_path ) ;
			char *server_name = next_word( "server name" ) ;
			char *server ;		/* for xinetd config file */

			if ( EQ( server_of_server_path, TCPD_NAME ) )
			{
				if ( ! uses_tcpd )
				{
					Sprint( 2, "You must use option %s if you use %s\n",
						DAEMON_DIR_OPTION, TCPD_NAME ) ;
					exit( 1 ) ;
				}
				if ( server_name[ 0 ] == '/' )
					server = server_name ;
				else
					server = make_pathname( 2, 
						daemon_dirpath, server_name ) ;
			}
			else
				server = server_path ;

			print_line( "server", server ) ;

			word = str_component( strp ) ;	/* 1st arg */
			if ( word != NULL )
			{
				Sprint( 1, "\t%-*s = %s", FIELD_WIDTH, 
					"server_args", word ) ;
				while ( (word = str_component( strp )) )
					Sprint( 1, " %s", word ) ;
				Sputchar( 1, '\n' ) ;
			}
		}

		Sprint( 1, "}\n\n" ) ;
	}
	Sflush( 1 ) ;
	exit( 0 ) ;
}


static void print_line( const char *name, const char *value )
{
	Sprint( 1, "\t%-*s = %s\n", FIELD_WIDTH, name, value ) ;
}


static char *next_word( const char *description )
{
	char *word = str_component( strp ) ;

	if ( word == NULL )
	{
		Sprint( 2, "Line %d: %s missing \n", line_count, description ) ;
		exit( 1 ) ;
	}
	return( word ) ;
}

static char *make_string_cat( const unsigned count, ... )
{
   va_list ap ;
   register unsigned i ;
   register unsigned len = 0 ;
   register char *s, *p ;
   char *newstring ;

   if ( count == 0 )
      return( NULL ) ;

   va_start( ap, count ) ;
   if (count == 1)
   {  /* 9 out of 10 have just 1, so this optimizes it */
      s = va_arg( ap, char * ) ;
      va_end( ap );
      if ( s == NULL )
         return strdup("");
      else
         return strdup(s);
   }
   for ( i = 0 ; i < count ; i++ )
   {
      s = va_arg( ap, char * ) ;
      if ( s == NULL )
         continue ;
      len += strlen( s ) ;
   }
   va_end( ap ) ;

   newstring = (char *)malloc( len + 1 ) ;
   if ( newstring == NULL )
      return( NULL ) ;

   p = newstring ;
   va_start( ap, count ) ;
   for ( i = 0 ; i < count ; i++ )
   {
      s = va_arg( ap, char * ) ;
      if ( s == NULL )
         continue ;
      while ( (*p++ = *s++) ) ;
      p-- ;
   }
   va_end( ap ) ;
   newstring[len] = 0;	/* if len == 0, must terminate or boom! */
   return newstring ;
}

static char *make_pathname( const unsigned count, ... )
{
   va_list ap ;
   register unsigned i ;
   register unsigned len = 0 ;
   register char *s, *p ;
   char *pathname ;

   if ( count == 0 )
      return( NULL ) ;

   va_start( ap, count ) ;
   for ( i = 0 ; i < count ; i++ )
   {
      s = va_arg( ap, char * ) ;
      len += strlen( s ) ;
   }
   va_end( ap ) ;

   pathname = (char *)malloc( len + count ) ;
   if ( pathname == NULL )
      return( NULL ) ;

   p = pathname ;
   va_start( ap, count ) ;
   for ( i = 0 ; i < count ; i++ )
   {
      s = va_arg( ap, char * ) ;
      while ( (*p++ = *s++) ) ;
      *(p-1) = '/' ;         /* change '\0' to '/' */
   }
   *(p-1) = '\0' ;
   va_end( ap ) ;
   return( pathname ) ;
} 

